zfit/core/loss.py

```diff
@@ -1,25 +1,220 @@
+
+    @property
+    # @abc.abstractmethod
+    def fit_range(self):
+        raise NotImplementedError
+
+
+class BaseLoss(LossInterface):
+
+    def __init__(self, pdf, data, fit_range, constraints=None):
+        # constraints = {} if constraints == {} else constraints # protect mutable argument
+        self._simultaneous = None
+        if constraints is None:
+            constraints = {}
+        pdf, data, fit_range = self._input_check(pdf, data, fit_range)
+        self._pdf = pdf
+        self._data = data
+        self._fit_range = fit_range
+        self._constraints = constraints
+
+    def __init_subclass__(cls, **kwargs):
+        cls._name = "UnnamedSubBaseLoss"
+
+    def _input_check(self, pdf, data, fit_range):
+        # TODO
+        return pdf, data, fit_range
+
+    def add_constraint(self, constraint):
+        if not isinstance(constraint, dict):
+            raise TypeError("`constraint` has to be a dict, is currently {}".format(type(constraint)))
+        overwritting_keys = set(constraint).intersection(self._constraints)
+        if overwritting_keys:
+            raise ValueError("Cannot change existing constraints but only add (currently). Constrain for "
+                             "parameter(s) {} already there.".format(overwritting_keys))
+        self._constraints.update(constraint)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def pdf(self):
+        return self._pdf
+
+    @property
+    def data(self):
+        return self._data
+
+    @property
+    def fit_range(self):
+        fit_range = self._fit_range
+        # fit_range = convert_to_range(self._fit_range)
+        return fit_range
+
+    @property
+    def constraints(self):
+        return self._constraints
+
+    @abc.abstractmethod
+    def _loss_func(self, pdf, data, fit_range, constraints=None):
+        raise NotImplementedError
+
+    def eval(self):
+        try:
+            return self._loss_func(pdf=self.pdf, data=self.data, fit_range=self.fit_range, constraints=self.constraints)
+        except NotImplementedError:
+            raise NotImplementedError("_loss_func not defined!")
+
+
+def errordef_nll(sigma):
+    return sigma
+
+
+class UnbinnedNLL(BaseLoss):
+    _name = "UnbinnedNLL"
+
+    def _loss_func(self, pdf, data, fit_range, constraints):
+        return unbinned_nll(pdf=pdf, data=data, fit_range=fit_range, constraints=constraints)
+
+    def errordef(self, sigma):
+        return errordef_nll(sigma)
+
+
+class SimpleLoss(BaseLoss):
+    _name = "SimpleLoss"
+
+    def __init__(self, func):
+        self._simple_func = func
+        super().__init__(pdf=None, data=None, fit_range=None)
+
+    def errordef(self, func):
+        raise NotImplementedError("For this simple loss function, no error calculation is possible.")
+
+    def _loss_func(self, pdf, data, fit_range, constraints=None):
+        loss = self._simple_func
+        return loss()
+
+
+def make_loss(func):
```

Comment:
It's a short-cut I've added, let's see if needed. But while the loss object is very useful for error estimation, it is actually not needed to just do a minimization. E.g. if you just build a TF function and want it to be minimized, `make_loss` simply converts that to a trivial loss function that can be minimized. And raises errors if it tries to go beyond the borders like calculating errors.

I am not sure whether this should be there (or be high level) but it's at least very useful for testing (though then it could be moved to the tests) and probably to do a simple minimization. Let's say, it's a wrapper around a simple TF Tensor that can be used for minimization.

More of a side-product. Don't like it?

Reply:
I was wondering why not simply do `minimize(SimpleLoss(x))`?

Reply:
My idea was to probably use it inside a pdf as an argument converter if it's a Tensor. But I don't like the idea too much. So it's just a little factory. But no real need. Remove?

Reply:
I would say remove for the time being.